### ISSUE TITLE
[JS Renderes][Accessibility] [Focus Lost for ToggleVisibility]

### ISSUE THREAD

**Issue by niravk-microsoft** [microsoft/AdaptiveCards#4949]
# JS Renderer

We are using JSRenderer to render Adaptive Cards into website as well as we are sending Adaptive Cards to Outlook. We are also going to send same adaptive cards to Teams as well

# Version of JS Adaptive cards is 2.4.0

# Details
We are using Toggle Visibility to achieve expand and collapse functionality. Now when we try to use tab key to reach to expand image and then press key to expand, focus is getting lost. However expected result is to set focus to toggled visible control i.e. collapsed image in this case.

It will be great help if we can get suggestion on how we can set focus of collapse image once we expand using key pressed event.

**Comment by shalinijoshi19** [comment]
@dclaux possible for you to take a first pass at this and feasibility of a fix/workaround?

**Comment by dclaux** [comment]
@shalinijoshi19 @niravk-microsoft There is no easy fix to this. We have done work to lose focus if the element being clicked (i.e. the element that triggers to ToggleVisibility) remains displayed after being clicked, but it is not the case in this scenario, where the "expand" button is made invisible and a "collapse" button is made visible.

The only potential solution I can think of would be to:
- If the element being clicked on is still visible after ToggleVisibility has executed, focus stays on it
- Otherwise, pick the first focusable visible element from the list of elements in the `targetElements` collection and give it focus
- If there is no such element, pick the first focusable visible element in the card and give it focus
- If there is no such element, focus is lost

If this is an acceptable algorithm, we will have to implement it across all renders. Note that it will not always produce the desired result.

The alternative to this solution is to introduce a new property in the schema to allow the card author to specify which element should have focus on completion of the ToggleVisibility action. But that would be a new feature, not a bug.

**Comment by niravk-microsoft** [comment]
I think it will be great if we can try two phased approach i.e. start with the fix for potential solution and then later add feature as an alternative solution for the same. In that way, current accessibility bug gets fixed and in future as a part of feature, we are giving control to Host of card to decide where to set focus on toggle visibility.

**Comment by shalinijoshi19** [comment]
@niravk-microsoft even the first solution is something that is new unplanned work and effectively will need to be discussed and then estimated before we can get you a timeline for this.. Can you help us understand the relative priority of this bug for you?

**Comment by RebeccaAnne** [comment]
Toggle accessibility is inherently impossible to solve fully for accessibility. We should look into options here to provide a better experience in these scenarios.

**Comment by dclaux** [comment]
I suggest simply adding a `moveFocusTo` optional property that can be set to a prioritized array of element ids. After the ToggleVisibility is executed, the focus is moved to the first focusable element in the array.

**Example:**
The below card switches from showing "input1" and "input2". With the specified `moveFocusTo` property, focus is moved to "input1" when "input2" is hidden and the other way around.
```
{
    "type": "AdaptiveCard",
    "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
    "version": "1.3",
    "body": [
        {
            "type": "Input.Text",
            "placeholder": "Placeholder text",
            "id": "input1",
            "label": "Input 1"
        },
        {
            "type": "Input.Text",
            "placeholder": "Placeholder text",
            "id": "input2",
            "label": "Input 2",
            "isVisible": false
        }
    ],
    "actions": [
        {
            "type": "Action.ToggleVisibility",
            "title": "Tottle between input 1 and input 2",
            "targetElements": [ "input1", "input2" ],
            "moveFocusTo": [ "input1", "input2" ]
        }
    ]
}
```

**Comment by jonmill** [comment]
We're getting conflicting info from different accessibility folks; closing until we can close on the right direction
